Linear Algebra Bibliography:
* https://opencourses.auth.gr/modules/units/?course=OCRS111&id=1289
* http://ebooks.edu.gr/ebooks/v/html/8547/2732/Mathimatika-G-Lykeiou-ThSp_html-apli/indexA1_6.html
* http://ecourse.uoi.gr/course/view.php?id=1225&lang=en
* http://ecourses.dbnet.ntua.gr/14011.html


\begin{array}{lcl} z & = & a \\ f(x,y,z) & = & x + 3y + z \end{array}

\begin{array}{rr} 2x + 3y + 8z & = & 5 \\ 2x - 7y + 3z & = & 6 \\ 4x + 3y + 21z & = & 32 \\ 5x + 6y - 9z & = & 5 \end{array}

\begin{array}{rrrr} 2*x & + & 3*y & + & 8*z & = & 5 \\ 2*x & - & 7*y & + & 3*z & = & 6 \\ 4*x & + & 3*y & + & 21*z & = & 32 \\ 5*x & + & 6*y & - & 9*z & = & 5 \end{array}

## System of equations

\begin{array}{rrrr} 2x & + & 3y & + & 8z & = & 5 \\ 2x & - & 7y & + & 3z & = & 6 \\ 4x & + & 3y & + & 21z & = & 32 \\ 5x & + & 6y & - & 9z & = & 5 \end{array}

In [223]:
import numpy as np


A = [[2,3,8,5], [2, -7 , 3, 6], [4, 3, 21, 32], [5, 6, -9, 5]]
A = np.array(A, dtype=np.float32)
A_orig = A.copy()

In [224]:
A[0, :] /= A[0, 0]
for i in range(1,4):
    A[i, :] -= A[0, :] * A[i, 0]
A

array([[  1. ,   1.5,   4. ,   2.5],
       [  0. , -10. ,  -5. ,   1. ],
       [  0. ,  -3. ,   5. ,  22. ],
       [  0. ,  -1.5, -29. ,  -7.5]], dtype=float32)

In [225]:
A[1, :] = A[1, :]/A[1, 1]
for i in range(2,4):
    A[i, :] -= A[1, :] * A[i, 1]
A

array([[  1.  ,   1.5 ,   4.  ,   2.5 ],
       [ -0.  ,   1.  ,   0.5 ,  -0.1 ],
       [  0.  ,   0.  ,   6.5 ,  21.7 ],
       [  0.  ,   0.  , -28.25,  -7.65]], dtype=float32)

In [197]:
A[2, :] = A[2, :]/A[2, 2]
for i in range(3,4):
    A[i, :] -= A[2, :] * A[i, 2]
A

array([[ 1.       ,  1.5      ,  4.       ,  2.5      ],
       [-0.       ,  1.       ,  0.5      , -0.1      ],
       [ 0.       ,  0.       ,  1.       ,  3.3384616],
       [ 0.       ,  0.       ,  0.       , 86.66154  ]], dtype=float32)

In [ ]:
array([[ 1.       ,  1.5      ,  4.       ,  2.5      ],
       [-0.       ,  1.       ,  0.5      , -0.1      ],
       [ 0.       ,  0.       ,  1.       ,  3.3384616

In [198]:
for i in range(1,-1,-1):
    A[i, :] -= A[2, :] * A[i, 2]
A

array([[  1.       ,   1.5      ,   0.       , -10.853847 ],
       [ -0.       ,   1.       ,   0.       ,  -1.7692308],
       [  0.       ,   0.       ,   1.       ,   3.3384616],
       [  0.       ,   0.       ,   0.       ,  86.66154  ]],
      dtype=float32)

In [226]:
for i in range(0,-1,-1):
    A[i, :] -= A[1, :] * A[i, 1]
A

array([[  1.  ,   0.  ,   3.25,   2.65],
       [ -0.  ,   1.  ,   0.5 ,  -0.1 ],
       [  0.  ,   0.  ,   6.5 ,  21.7 ],
       [  0.  ,   0.  , -28.25,  -7.65]], dtype=float32)

In [206]:
A_orig[:3,:3]@A[:3, 3]

array([ 4.999998,  5.999999, 32.      ], dtype=float32)

In [130]:
A_orig


array([[ 2.,  3.,  8.,  5.],
       [ 2., -7.,  3.,  6.],
       [ 4.,  3., 21., 32.],
       [ 5.,  6., -9.,  5.]], dtype=float32)

In [176]:
import pdb

In [244]:
def swap_rows(A, k, l):
    A[[k,l], :] = A[[l, k], :]


def solve_with_gaussian_elimination(A, contains_b=True):
    """
    Based on pseudocode from wikipedia.
    :param contains_b: if True then the last column of A is vector b from the system of equations Ax = b
    """
    assert A.ndim == 2
    
    A = A.copy() 
    
    m, n =  A.shape[:2]
    r = 0 # /* Initialization of the pivot row */
    c = 0 # /* Initialization of the pivot column */
    
    pivot_locations = [] ## store the locations of all pivot elements
    
    n_limit = n if not contains_b else n-1   ## if the last column is b then don't go there!
    
    while (r < m and c < n_limit): 
        # /* Find the k-th pivot: */
        r_max = np.argmax(A[r:, c]) + r
#         pdb.set_trace()
        if A[r_max, c] == 0:
            #/* No pivot in this column, pass to next column */
            c += 1 
        else:
#             swap_rows(A, r, r_max)
            # /* Do for all rows below pivot: */
            
            A[r, :] /= A[r, c]  ## make first element equal to one
            pivot_locations.append([r,c])
            
            f = A[r+1:, c:c+1] # / A[h, c]
            
            A[r+1:, :] = A[r+1:, :]  - f * A[r:r+1, :] 
            
            # /* Fill with zeros the lower part of pivot column: */
            A[r+1:, c] = 0 
            
            # /* Increase pivot row and column */
            r += 1 
            c += 1
    
    ## go backwards
    for pivot_loc in pivot_locations[:0:-1]:  ## we don't need the first row
        r, c = pivot_loc
        f = A[:r, c:c+1]
#         print(r,c)
#         pdb.set_trace()
        A[:r, c:] = A[:r, c:] - f * A[r:r+1, c:]  
        # /* Fill with zeros the upper part of pivot column: */
        A[:r, c] = 0 
        
    return A


In [245]:
B_orig = np.array([[1, 2, -1, 5, 6, 10],
                   [-1, -2, 1, -3, -2, -4],
                   [2, 4, -2, 6, 3, 6]], dtype=np.float32)
solve_with_gaussian_elimination(B_orig)

array([[ 1.,  2., -1.,  0.,  0.,  3.],
       [ 0.,  0.,  0.,  1.,  0., -1.],
       [-0., -0., -0., -0.,  1.,  2.]], dtype=float32)

In [246]:
C_orig = np.array([[1, -1, 2, -2, 3, 2],
                   [3, -3, -1, 1, 1, 5],
                   [-2, 2, 3, -3, 1, -4]], dtype=np.float32)
solve_with_gaussian_elimination(C_orig)

array([[ 1., -1.,  0.,  0.,  0.,  1.],
       [-0., -0.,  1., -1.,  0., -1.],
       [-0., -0., -0., -0.,  1.,  1.]], dtype=float32)

In [232]:
solve_with_gaussian_elimination(A_orig[:3, :], contains_b=True)

array([[ 1.       ,  0.       ,  0.       , -8.200001 ],
       [-0.       ,  1.       ,  0.       , -1.7692308],
       [ 0.       ,  0.       ,  1.       ,  3.3384616]], dtype=float32)

## Solving the system using Gaussian elimination: 

**with 2 equations for 3 unkowns:**
\begin{array}{rrrr} 2x & + & 3y & + & 8z & = & 5 \\ 2x & - & 7y & + & 3z & = & 6 \end{array}
returns: 

In [233]:
solve_with_gaussian_elimination(A_orig[:2, :], contains_b=True)

array([[ 1.  ,  0.  ,  3.25,  2.65],
       [-0.  ,  1.  ,  0.5 , -0.1 ]], dtype=float32)

which has as a solution the line:
\begin{array}{rrrr} x & = & 2.65 - 3t \\ y & = & -0.1 - 0.5t  \\ z & = & t \end{array}
where $t \in \mathbb{R}$

**with 3 equations for 3 unkowns:**: 
\begin{array}{rrrr} 2x & + & 3y & + & 8z & = & 5 \\ 2x & - & 7y & + & 3z & = & 6 \\ 4x & + & 3y & + & 21z & = & 32 \end{array}
returns: 

In [234]:
solve_with_gaussian_elimination(A_orig[:3, :], contains_b=True)

array([[ 1.       ,  0.       ,  0.       , -8.200001 ],
       [-0.       ,  1.       ,  0.       , -1.7692308],
       [ 0.       ,  0.       ,  1.       ,  3.3384616]], dtype=float32)

which is a unique solution (x, y, z) = (-8.2, -1.77, 3.34)

3. with 4 equations for 3 unkowns: \begin{array}{rrrr} 2x & + & 3y & + & 8z & = & 5 \\ 2x & - & 7y & + & 3z & = & 6 \\ 4x & + & 3y & + & 21z & = & 32 \\ 5x & + & 6y & - & 9z & = & 5 \end{array}
   returns: 

In [235]:
solve_with_gaussian_elimination(A_orig[:4, :], contains_b=True)

array([[ 1.       ,  0.       ,  0.       , -8.200001 ],
       [-0.       ,  1.       ,  0.       , -1.7692308],
       [ 0.       ,  0.       ,  1.       ,  3.3384616],
       [ 0.       ,  0.       ,  0.       , 86.66154  ]], dtype=float32)

## Using normal equation

In [107]:
A_orig
A, b = A_orig[:, :3], A_orig[:, 3:]
A, b

(array([[ 2.,  3.,  8.],
        [ 2., -7.,  3.],
        [ 4.,  3., 21.],
        [ 5.,  6., -9.]], dtype=float32), array([[ 5.],
        [ 6.],
        [32.],
        [ 5.]], dtype=float32))

In [117]:
def solve(A, b):
    print(A, b)
    pseudo_inv = np.linalg.pinv(A)
    return pseudo_inv @ b

In [111]:
solve(A, b)

array([[2.403015  ],
       [0.06265867],
       [0.90363723]], dtype=float32)

In [ ]:
np.linalg.inv(A[:3, :])

In [118]:
solve(A[:3, :], b[:3, :])

[[ 2.  3.  8.]
 [ 2. -7.  3.]
 [ 4.  3. 21.]] [[ 5.]
 [ 6.]
 [32.]]


array([[-8.200002 ],
       [-1.7692308],
       [ 3.3384616]], dtype=float32)

### With an undertermined system (only 2 equations for 3 dimensions) 
we get a point on the line in 3D spac that has all possible solutions x,y,z

In [114]:
solve(A[:2, :], b[:2, :])

array([[ 0.2941799 ],
       [-0.46243384],
       [ 0.7248677 ]], dtype=float32)

In [115]:
t = 0.7248677
x = 2.65 - 3.25*t
y = -0.1 -0.5*t
z = t 
x, y ,z

(0.29417997500000004, -0.46243385000000004, 0.7248677)